### DISCUSSION TITLE
[PB as Framework] custom route testing broken in 0.17

### DISCUSSION THREAD

**Discussion by pnmcosta** [pocketbase/pocketbase#3025] [Q&A] [answered]
Great job on the new release, I do have some custom route tests breaking.

Easy to replicate, I had a quick look but couldn't figure out why yet, the code below works fine on 0.16.x, but on 0.17.x it always 404's:

```
func BindHandlers(app core.App) {
	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.GET("/test", func(c echo.Context) error {
			return c.NoContent(http.StatusOK)
		})

		return nil
	})
}

func main() {
	app := pocketbase.New()

	BindHandlers(app)
	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}

func TestMain(t *testing.T) {
	scenarios := []tests.ApiScenario{
		{
			Name:            "should 200",
			Method:          http.MethodGet,
			Url:             "/test",
			ExpectedStatus:  http.StatusOK,
			ExpectedContent: []string{""},
			TestAppFactory: func() (*tests.TestApp, error) {
				app, err := tests.NewTestApp()
				if err != nil {
					return nil, err
				}

				BindHandlers(app)
				return app, nil
			},
		},
	}

	for _, scenario := range scenarios {
		scenario.Test(t)
	}
}
```

It does however `go run main.go serve` both fine as expected.

**Comment by ganigeorgiev** [comment, marked as answer]
Hm, I think this is because I moved the `OnBeforeServe` trigger inside `apis.Serve` which is not called by the `ApiScenario`.

I'll have a more detailed look at it tomorrow during the day, but if you want a temporary workaround you can try to manually trigger the hook with the minimal event data inside the `ApiScenario.BeforeTestFunc`:

```go
scenarios := []tests.ApiScenario{
	{
		Name:            "should 200",
		Method:          http.MethodGet,
		Url:             "/test",
		ExpectedStatus:  http.StatusOK,
		ExpectedContent: []string{""},
		TestAppFactory: func() (*tests.TestApp, error) {
			app, err := tests.NewTestApp()
			if err != nil {
				return nil, err
			}

			BindHandlers(app)

			return app, nil
		},
		BeforeTestFunc: func(t *testing.T, app *tests.TestApp, e *echo.Echo) {
			event := &core.ServeEvent{
				App:    app,
				Router: e,
			}
			app.OnBeforeServe().Trigger(event)
		},
	},
}
```

**Comment by pnmcosta** [comment]
Thank you, figured something like that was the case. This fixes it 🙇 

I have updated my fork with this patch as I have many tests and updating them all would be too much 🤣 

I also would like to point you towards another change I did on my fork and hopefully you can consider it.

The `TestAppFactory` function accepting the testing pointer: https://github.com/pocketbase/pocketbase/commit/ea9d960b3e05da8e2cee8caa38c0688d2b2ea56d

This is because I use adapter mocks that I need to inject to the test app and expect/return. I'm using `gomock`.

If I don't create the mock controller with the test created by the run: https://github.com/pocketbase/pocketbase/blob/master/tests/api.go#L54 any mock failures prevent sibling test runs to complete (e.g. it's the parent test that gets failure rather than the specific run).

Thank you

**Reply by ganigeorgiev** [reply]
> This is because I use adapter mocks that I need to inject to the test app and expect/return. 

Could you provide a code sample or example showing how do you use the testing argument?

We can pass the testing instance but it may cause confusion to the `TestAppFactory` because currently its declaration expect returning an error on failure and it may not be clear for the user whether they need to manually trigger a test failure (eg. with `t.Fatal(err)`) or return the error.

**Reply by pnmcosta** [reply]
Understood, yeah it's not very idiomatic :/ but it was the only way I figured out to be able to do it because I need some adapters at serve time (couldn't use the `BeforeTestFunc` it's too late).

This [gist](https://gist.github.com/pnmcosta/c40357cc3bccef8e92b79aea75a9286f) (multiple files) is a trimmed version of what I'm doing, please note it's replacing PB go.mod with my fork: `replace github.com/pocketbase/pocketbase => github.com/pnmcosta/pocketbase v0.0.0-20230731075920-4239dd427f2c`

**Reply by ganigeorgiev** [reply]
In this case, Yes maybe it make sense to have the testing instance in the factory.

I've added a note in `master` to consider changing the declaration and remove the return error of the `ApiScenario.TestAppFactory` hook in the future but for now I want to avoid introducing breaking changes in minor releases so this will be left for v0.18.0+ (_there are also plans to change `ApiScenario.AfterTestFunc` but I haven't got the time to think properly on it yet_).

**Reply by pnmcosta** [reply]
Awesome, looking forward to it.

**Reply by ganigeorgiev** [reply]
Just for info in case someone stumble here, I've pused in the [`develop` branch](https://github.com/pocketbase/pocketbase/commit/311bc74b7e971bdf88cdc0deece009f146d529e3) 2 small changes to the `tests.ApiScenario` struct:

- The `ApiScenario.AfterTestFunc` now receive as 3rd argument `*http.Response` pointer instead of `*echo.Echo` as the latter is not really useful in this context.
  ```go
  // old
  AfterTestFunc: func(t *testing.T, app *tests.TestApp, e *echo.Echo)

  // new
  AfterTestFunc: func(t *testing.T, app *tests.TestApp, res *http.Response)
  ```

- The `ApiScenario.TestAppFactory` now accept the test instance as argument and no longer expect an error as return result ([#3025](https://github.com/pocketbase/pocketbase/discussions/3025#discussioncomment-6592272)).
  ```go
  // old
  TestAppFactory: func() (*tests.TestApp, error)

  // new
  TestAppFactory: func(t *testing.T) *tests.TestApp
  ```
  Returning a `nil` app instance from the factory results in test failure. You can enforce a custom failure error by calling `t.Fatal(err)` inside the factory.


_The above changes will be available with the next v0.18.0 release (there are no ETAs yet)._
